table-api/src/main/java/io/deephaven/api/ConcurrencyControl.java

```diff
@@ -0,0 +1,35 @@
+//
+// Copyright (c) 2016-2025 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.api;
+
+import io.deephaven.api.filter.Filter;
+
+/**
+ * Provides concurrency control operations for expressions such as {@link Filter filters} and {@link Selectable
+ * selectables}.
+ * <p>
+ * This interface allows the application of different concurrency strategies to an expression. Through the provided
+ * methods, clients can enforce serial execution, apply barrier constraints, and specify ordering of operations relative
+ * to barriers. This facilitates both intra- and inter-expression synchronization, ensuring that expressions can be
+ * executed without risking concurrent modifications or out-of-order processing when interacting with stateful parts of
+ * custom user code.
+ *
+ * @param <T> the type of the expression to which concurrency control is applied.
+ */
+public interface ConcurrencyControl<T> {
+
+    /**
+     * Applies serial concurrency control to the expression.
+     * <p>
+     * <ul>
+     * <li>Concurrency impact: The expression will never be invoked concurrently with itself.</li>
+     * <li>Intra-expression ordering impact: Rows are evaluated sequentially in row set order.</li>
+     * <li>Inter-expression ordering impact: Acts as an absolute reordering barrier, ensuring that no parts of a where
```

Comment:
It is not clear to me that there should be any inter-expression ordering impact for serial.  If you wanted the barrier, you could express it as a barrier; but if you have two expressions that are both serial there is no reason we could not at some point choose to evaluate them on separate threads.

Reply:
Chatted offline; added the paragraph you suggested that more clearly discusses behavior. Leaving this note as-is per discussion.